src/main/scala/viper/gobra/util/ViperChopper.scala

```diff
@@ -62,27 +62,56 @@ object ViperChopper {
+          val t2 = System.nanoTime()
+          val mergedPrograms = mergePrograms(smallestPrograms)(bound, penalty.contravariantLift(idToVertex))
+          val res = mergedPrograms.map(_.toSet)
+          val t3 = System.nanoTime()
+          timeCutting = BigDecimal((t2 - t1) / 1e9d).setScale(2, BigDecimal.RoundingMode.HALF_UP)
+          timeMerging = BigDecimal((t3 - t2) / 1e9d).setScale(2, BigDecimal.RoundingMode.HALF_UP)
+          maxNumberOfParts = smallestPrograms.size
+          forallSmallNode = f => smallestPrograms.forall(_.forall(f))
+          res
+        } else {
+          val t0 = System.nanoTime()
+          val (_, idToSCC, sccEdges) = SCC.fastCompute(N, edges)
+          val sccNodes = nodes.map(idToSCC) // may contain duplicates, but smallestCut can deal with that
+          val t1 = System.nanoTime()
+          val smallestPrograms = smallestCut(N, sccNodes, sccEdges, (x: SCC.Component[Int]) => x.proxy)
+          val t2 = System.nanoTime()
+          val mergedPrograms = mergePrograms(smallestPrograms)(bound, penalty.contravariantSumLift(_.nodes.map(idToVertex)))
+          val res = mergedPrograms.map(_.toSet.flatMap((c: SCC.Component[Int]) => c.nodes))
+          val t3 = System.nanoTime()
+          timeSCC     = BigDecimal((t1 - t0) / 1e9d).setScale(2, BigDecimal.RoundingMode.HALF_UP)
+          timeCutting = BigDecimal((t2 - t1) / 1e9d).setScale(2, BigDecimal.RoundingMode.HALF_UP)
+          timeMerging = BigDecimal((t3 - t2) / 1e9d).setScale(2, BigDecimal.RoundingMode.HALF_UP)
+          maxNumberOfParts = smallestPrograms.size
+          forallSmallNode = f => smallestPrograms.forall(_.forall(_.nodes.forall(f)))
+          res
+        }
+      }
 
-      println(s"Chopped verification condition into ${result.size} parts. Maximum number of parts is ${smallestPrograms.size}. (Time: ${timeCutting}s + ${timeMerging}s)")
+      println(s"Chopped verification condition into ${result.size} parts. Maximum number of parts is $maxNumberOfParts. (Time: ${timeSCC}s + ${timeCutting}s + ${timeMerging}s)")
 
       // Safety check validating the result partially. Can be removed in a year if no error has been found.
       {
         val containedInResult = Array.ofDim[Boolean](N)
         for (program <- result; node <- program) { containedInResult(node) = true }
         // check all nodes of the smallest programs are present (no node should be lost)
-        val smallestProgramContainedInResult = smallestPrograms.forall(_.forall(containedInResult(_)))
+        val smallestProgramContainedInResult = forallSmallNode(containedInResult(_))
         assert(smallestProgramContainedInResult, "Chopper Error: Lost nodes during merging step.")
 
         val containedInSmallest = Array.ofDim[Boolean](N)
-        for (program <- smallestPrograms; node <- program) { containedInSmallest(node) = true }
+        forallSmallNode{node => containedInSmallest(node) = true; true }
```

Comment:
This looks somewhat hacky, but it is fine, given that this code is supposed to be deleted eventually